pkg/extensions/tbtc/tbtc.go

```diff
@@ -97,6 +134,270 @@ func (t *tbtc) monitorRetrievePubKey(
+		redeemedSubscription, err := t.chain.OnDepositRedeemed(handler)
+		if err != nil {
+			return nil, err
+		}
+
+		return subscription.NewEventSubscription(
+			func() {
+				signatureSubscription.Unsubscribe()
+				redeemedSubscription.Unsubscribe()
+			},
+		), nil
+	}
+
+	actFn := func(depositAddress string) error {
+		keepAddress, err := t.chain.KeepAddress(depositAddress)
+		if err != nil {
+			return err
+		}
+
+		redemptionRequestedEvents, err := t.chain.PastDepositRedemptionRequestedEvents(
+			depositAddress,
+			0, // TODO: Should be something better that 0
+		)
+		if err != nil {
+			return err
+		}
+
+		if len(redemptionRequestedEvents) == 0 {
+			return fmt.Errorf(
+				"no redemption requested events found for deposit: [%v]",
+				depositAddress,
+			)
+		}
+
+		latestRedemptionRequestedEvent :=
+			redemptionRequestedEvents[len(redemptionRequestedEvents)-1]
+
+		signatureSubmittedEvents, err := t.chain.PastSignatureSubmittedEvents(
+			keepAddress,
+			latestRedemptionRequestedEvent.BlockNumber,
+		)
+		if err != nil {
+			return err
+		}
+
+		depositDigest := latestRedemptionRequestedEvent.Digest
+
+		for _, signatureSubmittedEvent := range signatureSubmittedEvents {
+			if bytes.Equal(signatureSubmittedEvent.Digest[:], depositDigest[:]) {
+				// We add 27 to the recovery ID to align it with ethereum and
+				// bitcoin protocols where 27 is added to recovery ID to
+				// indicate usage of uncompressed public keys.
+				v := 27 + signatureSubmittedEvent.RecoveryID
+
+				return t.chain.ProvideRedemptionSignature(
+					depositAddress,
+					v,
+					signatureSubmittedEvent.R,
+					signatureSubmittedEvent.S,
+				)
+			}
+		}
+
+		return fmt.Errorf(
+			"could not find signature for digest: [%v]",
+			depositDigest,
+		)
+	}
+
+	monitoringSubscription, err := t.monitorAndAct(
+		ctx,
+		"provide redemption signature",
+		monitoringStartFn,
+		monitoringStopFn,
+		t.watchKeepClosed,
+		actFn,
+		actBackoffFn,
+		func(_ string) (time.Duration, error) {
+			return timeout, nil
+		},
+	)
+	if err != nil {
+		return err
+	}
+
+	go func() {
+		<-ctx.Done()
+		monitoringSubscription.Unsubscribe()
+		logger.Infof("provide redemption signature monitoring disabled")
+	}()
+
+	logger.Infof("provide redemption signature monitoring initialized")
+
+	return nil
+}
+
+func (t *tbtc) monitorProvideRedemptionProof(
+	ctx context.Context,
+	actBackoffFn backoffFn,
+	timeout time.Duration,
+) error {
+	monitoringStartFn := func(
+		handler depositEventHandler,
+	) (subscription.EventSubscription, error) {
+		// Start right after a redemption signature has been provided.
+		return t.chain.OnDepositGotRedemptionSignature(handler)
+	}
+
+	monitoringStopFn := func(
+		handler depositEventHandler,
+	) (subscription.EventSubscription, error) {
+		// Stop in case the redemption fee has been increased by someone else.
+		redemptionRequestedSubscription, err := t.chain.OnDepositRedemptionRequested(
```

Comment:
Here we not only need to confirm the redemption fee increase TX has enough confirmations (see previous comment) but also if the one we are looking at happened after TX in which the redemption signature has been provided.

Reply:
As I responded in the previous comment, I've added a TODO and I will handle this in a separate PR.

Reply:
Is this one already addressed in one of follow-up PRs?

Reply:
It's addressed in https://github.com/keep-network/keep-ecdsa/pull/592 where we check the status of deposit after `12` blocks.